OpenRA.Mods.Common/Traits/AutoTarget.cs

```diff
@@ -269,48 +269,67 @@ public Actor ScanForTarget(Actor self, bool allowMove)
+			if (target.Type != TargetType.Invalid)
+				Attack(self, target, allowMove);
 		}
 
-		void Attack(Actor self, Actor targetActor, bool allowMove)
+		void Attack(Actor self, Target target, bool allowMove)
 		{
-			TargetedActor = targetActor;
-			var target = Target.FromActor(targetActor);
 			self.SetTargetLine(target, Color.Red, false);
 
 			foreach (var ab in activeAttackBases)
 				ab.AttackTarget(target, false, allowMove);
 		}
 
-		Actor ChooseTarget(Actor self, AttackBase ab, Stance attackStances, WDist scanRange, bool allowMove)
+		Target ChooseTarget(Actor self, AttackBase ab, Stance attackStances, WDist scanRange, bool allowMove)
 		{
-			Actor chosenTarget = null;
+			var chosenTarget = Target.Invalid;
 			var chosenTargetPriority = int.MinValue;
 			int chosenTargetRange = 0;
 
 			var activePriorities = activeTargetPriorities.ToList();
 			if (activePriorities.Count == 0)
-				return null;
+				return chosenTarget;
 
-			var actorsInRange = self.World.FindActorsInCircle(self.CenterPosition, scanRange);
-			foreach (var actor in actorsInRange)
+			var targetsInRange = self.World.FindActorsInCircle(self.CenterPosition, scanRange)
+				.Select(Target.FromActor)
+				.Concat(self.Owner.FrozenActorLayer.FrozenActorsInCircle(self.World, self.CenterPosition, scanRange)
+					.Select(Target.FromFrozenActor));
+
+			foreach (var target in targetsInRange)
 			{
-				// PERF: Most units can only attack enemy units. If this is the case but the target is not an enemy, we
-				// can bail early and avoid the more expensive targeting checks and armament selection. For groups of
-				// allied units, this helps significantly reduce the cost of auto target scans. This is important as
-				// these groups will continuously rescan their allies until an enemy finally comes into range.
-				if (attackStances == OpenRA.Traits.Stance.Enemy && !actor.AppearsHostileTo(self))
+				BitSet<TargetableType> targetTypes;
+				if (target.Type == TargetType.Actor)
+				{
+					// PERF: Most units can only attack enemy units. If this is the case but the target is not an enemy, we
+					// can bail early and avoid the more expensive targeting checks and armament selection. For groups of
+					// allied units, this helps significantly reduce the cost of auto target scans. This is important as
+					// these groups will continuously rescan their allies until an enemy finally comes into range.
+					if (attackStances == OpenRA.Traits.Stance.Enemy && !target.Actor.AppearsHostileTo(self))
+						continue;
+
+					// Check whether we can auto-target this actor
+					targetTypes = target.Actor.GetEnabledTargetTypes();
+
+					if (PreventsAutoTarget(self, target.Actor) || !target.Actor.CanBeViewedByPlayer(self.Owner))
+						continue;
+				}
+				else if (target.Type == TargetType.FrozenActor)
+				{
+					if (attackStances == OpenRA.Traits.Stance.Enemy && self.Owner.Stances[target.FrozenActor.Owner] == OpenRA.Traits.Stance.Ally)
```

Comment:
How about a `Target.AppearsHostileTo()`?

Reply:
false, I guess. But it was just an off-the-cuff suggestion, so feel free to ignore it

Reply:
Ok. Lets leave it as is for now, then we can reevaluate this if we find a second usecase for such a method.